src/config_format/flb_cf_yaml.c

```diff
@@ -1338,114 +1408,252 @@ static char *get_real_path(char *file, char *path, size_t size)
-        if (!p) {
-            state_destroy(s);
-            return NULL;
-        }
-        s->caller_file = flb_sds_create(caller_file);
-        s->caller_root_path = flb_sds_create(caller_path);
+    if (last->allocation_flags & HAS_KEY) {
+        flb_sds_destroy(last->key);
     }
-    else {
-        s->caller_file = flb_sds_create(s->file);
-        s->caller_root_path = flb_sds_create(s->root_path);
+    if (last->allocation_flags & HAS_KEYVALS) {
+        cfl_kvlist_destroy(last->keyvals);
     }
+    state_destroy(last);
 
-    /* check if the file exists */
-    ret = stat(file, &st);
-    if (ret == 0) {
-        p = get_real_path(file, file_path, PATH_MAX + 1);
-        s->file = flb_sds_create(file);
-        s->root_path = flb_sds_create(file_path);
+    if (cfl_list_is_empty(&ctx->states)) {
+        return NULL;
     }
-    else if (errno == ENOENT && caller_file && s->caller_root_path != NULL) {
-        snprintf(file_path, PATH_MAX, "%s/%s", s->caller_root_path, file);
-        s->file = flb_sds_create(file_path);
+    return cfl_list_entry_last(&ctx->states, struct parser_state, _head);
+}
+
+static void state_destroy(struct parser_state *s)
+{
+    flb_free(s);
+}
+
+static struct parser_state *state_create(struct file_state *parent, struct file_state *file)
+{
+    struct parser_state *state;
+
+    /* allocate context */
+    state = flb_calloc(1, sizeof(struct parser_state));
+    if (!state) {
+        flb_errno();
+        return NULL;
+    }
+
+    state->file = file;
+#ifndef FLB_HAVE_STATIC_CONF
+    if (parent) {
+       state->file->parent = parent;
     }
+#else
+    s->file->name = flb_sds_create("***static***");
+    s->file->path = flb_sds_create("***static***");
 #endif
 
-    return s;
+    return state;
 }
 
 static int read_config(struct flb_cf *cf, struct local_ctx *ctx,
-                       char *caller_file, char *cfg_file)
+                       struct file_state *parent, char *cfg_file)
 {
     int ret;
     int status;
     int code = 0;
-    char *file;
     struct parser_state *state;
+    flb_sds_t file = NULL;
+    flb_sds_t include_file = NULL;
     yaml_parser_t parser;
     yaml_event_t event;
     FILE *fh;
+    struct file_state fstate;
+
+    if (parent && cfg_file[0] != '/') {
+        file = flb_sds_create_size(strlen(cfg_file) + strlen(parent->path));
+        if (file == NULL) {
+            flb_error("unable to create filename");
+            return -1;
+        }
+        flb_sds_printf(&file, "%s/%s", parent->path, cfg_file);
+    } else {
+        file = flb_sds_create(cfg_file);
+    }
 
-    state = state_create(caller_file, cfg_file);
+    include_file = flb_sds_create(file);
+    fstate.name = basename(file);
+    fstate.path = dirname(file);
+
+    fstate.parent = parent;
+
+    state = state_start(ctx, &fstate);
     if (!state) {
+        flb_error("unable to push initial include file state: %s", cfg_file);
         return -1;
```

Comment:
Release resources

Reply:
You are trying to release the argument named `cfg_file` instead of the local variable named `file` which you still need to rename.